cmd/kogito/command/deploy/deploy_service.go

```diff
@@ -278,3 +319,149 @@ func (i *deployCommand) Exec(cmd *cobra.Command, args []string) (err error) {
+	log := context.GetDefaultLogger()
+	options := &buildv1.BinaryBuildRequestOptions{}
+	var fileName string
+	// from url, do only basic check, for example, if the url has the suffix that's match the allowed ones
+	// if found download it and save on /tmp
+	switch {
+	case strings.HasPrefix(resource, "http"):
+		parsedURL, err := url.ParseRequestURI(resource)
+		if err != nil {
+			return nil, nil, fmt.Errorf("kogito resource is not a valid URL, received %s", resource)
+		}
+
+		hasExtension := false
+		ff := strings.Split(parsedURL.Path, "/")
+		fileName = strings.Join(strings.Fields(ff[len(ff)-1]), "")
+
+		if isSuffixSupported(fileName) {
+			hasExtension = true
+		}
+
+		if hasExtension {
+			response, err := http.Get(resource)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to download %s, error message: %s", resource, err.Error())
+			}
+			options.AsFile = fileName
+			return response.Body, options, nil
+		}
+
+	// handle single file, files on dir.
+	default:
+		fileInfo, err := os.Stat(resource)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if fileInfo.Mode().IsRegular() {
+			if isSuffixSupported(resource) {
+				log.Infof("File found %s.", resource)
+				ff := strings.Split(resource, "/")
+				fileName = strings.Join(strings.Fields(ff[len(ff)-1]), "")
+				fileReader, err := os.Open(resource)
+				if err != nil {
+					return nil, nil, err
+				}
+				options.AsFile = fileName
+				return fileReader, options, nil
+			}
+		}
+
+		if fileInfo.Mode().IsDir() {
+			log.Info("The provided source is a dir, packing files.")
+			if err != nil {
+				return nil, nil, err
+			}
+
+			var buf bytes.Buffer
+			gzipWriter := gzip.NewWriter(&buf)
+			tarWriter := tar.NewWriter(gzipWriter)
+
+			var link string
+			filepath.Walk(resource, func(absoluteFilePath string, fileInfo os.FileInfo, err error) error {
+				if isSuffixSupported(fileInfo.Name()) {
+
+					if fileInfo.Mode()&os.ModeSymlink != 0 {
+						link, err = os.Readlink(absoluteFilePath)
+						if err != nil {
+							return err
+						}
+					}
+					header, err := tar.FileInfoHeader(fileInfo, link)
+					if err != nil {
+						return err
+					}
+					header.Name = filepath.ToSlash(fileInfo.Name())
+					header.Linkname = filepath.ToSlash(header.Linkname)
+					header.Format = tar.FormatPAX
+
+					// write header
+					if err := tarWriter.WriteHeader(header); err != nil {
+						return err
+					}
+
+					fileToCompress, err := os.Open(absoluteFilePath)
+					if err != nil {
+						return err
+					}
+					defer fileToCompress.Close()
+
+					if _, err := io.Copy(tarWriter, fileToCompress); err != nil {
+						return err
+					}
+				}
+
+				return nil
+			})
+
+			// produce tar
+			if err := tarWriter.Close(); err != nil {
+				return nil, nil, err
+			}
+			// produce gzip
+			if err := gzipWriter.Close(); err != nil {
+				return nil, nil, err
+			}
+			tarWriter.Flush()
+
+			var r io.Reader
+			br := bufio.NewReaderSize(&buf, 4096)
+			r = br
```

Comment:
What is the purpose of this assignment, why not to use br directly?

Reply:
didn't work, if I return the buffer reader or the io.Reader it comes a empty file.
this was the only way I found so far to make it work.

Do you have any suggestion?

Reply:
My local experiments show that when I use br or &buf on https://github.com/kiegroup/kogito-cloud-operator/pull/247/files#diff-7b68eb431df193ad354067385d51880dR426 then content of the file is returned.
Checked using test:
```
func Test_Qqq(t *testing.T) {
	fileToWrite, _ := ioutil.TempFile(os.TempDir(), "compressed_kogito_resourcesxxx_*.tgz")
	reader, _, _ := matchKogitoResources("/home/ksuta/Downloads/qqq")
	io.Copy(fileToWrite, reader)
	fileToWrite.Sync()
	fileToWrite.Close()
}
```
(need to be careful as the test results can be cached, skipping execution in such case).

Reply:
When I think about it, what prevents us from returning directly &buf, skipping whole temp file handling?

Reply:
the content might be returned, but not on a correct tar.gz format, after build the buf only, without writing the file on filesyste, looks like it goes a empty file or not recognized one. I'll try again to double check it.

Reply:
Like I mentioned in other comment, please move all this file handling/compression to an util/compression package and deal with it there, so we can create a bunch of unit tests to verify all those behaviors.

Reply:
@sutaakar apparently it worked now, changed some things.